rfcs/0042-networking-refactor.md

```diff
@@ -0,0 +1,337 @@
+
+Below is an example of what the new gossip network interface would look like from the perspective of the rest of the daemon code. Note that it is much more abstract than before, modeling our new design choices regarding migrating state from OCaml to Go.
+
+```ocaml
+module Mina_net : sig
+  module Config : sig
+    type t = (* omitted *)
+  end
+
+  module Gossip_pipes : sig
+    type t =
+      { blocks: External_transition.t Strict_pipe.Reader.t
+      ; txn_pool_diffs: Transaction_pool.Diff.t Strict_pipe.Reader.t
+      ; snark_pool_diffs: Snark_pool.Diff.t Strict_pipe.Reader.t }
+  end
+
+  module Stats : sig
+    type t = (* omitted *)
+  end
+
+  module Abstract_peer_graph =
+    Graph.Persistent.Digraph.ConcreteBidirectional (struct
+      type t =
+        | AnyPeer
+        | SpecificPeer of
+      [@@deriving equal, hash]
+    end)
+
+  type t
+
+  (* We can construct the gossip network subsystem using the configuration,
+   * the set of RPC implementations, a state which is shared with the
+   * handlers of the provided RPC implementations. Once constructed, the
+   * gossip network handle will be returned along with pipes for reading
+   * incoming gossip network messages. *)
+  val create :
+       Config.t
+    -> ('state Rpc_intf.t_with_implementation) list
+    -> 'state
+    -> (t * Gossip_pipes.t) Deferred.Or_error.t
+
+  val stats : t -> Stats.t
+
+  (* Query operations now have the ability to express additional validation
+   * logic on a per-request basis, in addition to RPC-wide validation logic
+   * that is defined  *)
+  val query_peer :
+       t
+    -> Peer.t
+    -> ('q, 'r) Rpc_intf.rpc
+    -> 'q
+    -> ?f:('r Envelope.Incoming.t Deferred.t -> Validation_status.t Deferred.t)
+    -> 'r Envelope.Incoming.t Deferred.Or_error.t
+
+  val query_peers :
+       t
+    -> Abstract_peer_graph.t
+    -> ('q, 'r) Rpc_intf.rpc
+    -> 'q
+    -> ?f:('r Envelope.Incoming.t Deferred.t -> Validation_status.t Deferred.t)
+    -> 'r Envelope.Incoming.t Deferred.Or_error.t
+
+  val broadcast : t -> Gossip_message.t -> unit Deferred.t
+
+  val ban_peer : t -> Peer.t -> unit Deferred.t
+end
+```
+
+### Go Implementation
+
+The Go implementation will be fairly similar to how it's structured today. The scope of the state it maintains is more or less the same, the biggest changes introduced in this RFC effect the OCaml code more.  The main work in Go will just be switching it to use Cap'N Proto and use the new message format instead of the old one.
```

Comment:
if the go code is more or less the same, we should probably just remove the libp2p-peer-exchange option and accept fate that we have to just do the minaExchange manually